Support functions with parameters

iK4tsu opened the pull request:
Now functions with parameters and default argument values can be used with decorators.

iK4tsu committed:
feat: add D style varargs to the decorated function

The D style varargs make for a less strict a better usage of the
generated code. Other solutions involve using 'Parameters' to extract
the given function parameters, this would also remove the template
usage.

However, that brings problems with functions that have default
argument values. To have the same extact signature some magic with alias
sequences would need to be done.

That extra complexity doesn't bring any major value since we can
postpone this verification until when the actual function is called.

Signed-off-by: João Lourenço <jlourenco5691@gmail.com>

iK4tsu committed:
refactor: store the first impl call in a variable

Signed-off-by: João Lourenço <jlourenco5691@gmail.com>

iK4tsu committed:
feat: add support for functions with parameters

Signed-off-by: João Lourenço <jlourenco5691@gmail.com>

iK4tsu committed:
feat: correctly call functions with default argument values

Some functions have default argument values defined. To handle that
correctly we assume to pass (by order) the rest of the arguments the
user did not pass.

D doesn't have (yet) the support for named arguments. This means we can
assume the values are always passed by the same order they are defined
in their function signature. Assuming the user has following function:

void fun(int a = 4, int b) {}

The variable 'b' is mandatory, and to write to it, the user has to also
write to 'a' even though it is an optional argument. Knowing this, we
are safe to assume that the indices of the default values and the
quantity of actual arguments passed by the user comply with each other.

Note: void is used for any type that does not have a default value
resulting in an error, not the best error message but it's better than
nothing;

Signed-off-by: João Lourenço <jlourenco5691@gmail.com>

iK4tsu committed:
test: functions with default argument values

Signed-off-by: João Lourenço <jlourenco5691@gmail.com>

ljmf00 commented on source/decorated.d:
```diff
@@ -135,4 +143,38 @@ unittest
+
+        assert(fib() == 3); // can be used with default arguments
+    }
 }
```
add new line to conform with POSIX.

iK4tsu replied:
done

ljmf00 commented on source/decorated.d:
```diff
@@ -135,4 +143,38 @@ unittest
     }
+}
+
+unittest
```
Add `@safe`. Perhaps, can we add `@nogc` too? Is the introduced `fib` delegate guaranteed to be not dynamically allocated? We should also add a test with `pure` code to prove functionality.

iK4tsu replied:
Closures can never be @nogc, they are allocated with the GC

ljmf00 replied:
> Closures can never be @nogc, they are allocated with the GC

They can, as long as living in the stack and not escaped. We shouldn't enforce the creation of GC allocated functions.

iK4tsu replied:
I'm not enforcing anything??

ljmf00 replied:
> I'm not enforcing anything??

The test is doable with `@nogc`, according to the language semantics, but if isn't the case here, then your implementation is enforcing it to use the GC.

iK4tsu replied:
The test cannot be done with `@nogc`. The function makes use of `memoize` and returns a clojure based on it that gets allocated by the GC automatically.

ljmf00 commented on source/decorated.d:
```diff
@@ -135,4 +143,38 @@ unittest
+        @cache
+        mixin decorated!("fib", (size_t n = 1)
+        {
+            expensive++; // simulates an expensive algorithm
+            return expensive;
```
Why not actually adding fibonacci? Otherwise rename it and do the following change:

```suggestion
            return expensive++; // simulates an expensive algorithm
```

iK4tsu replied:
First, Fibonacci is not the best example to use here, and second, this does not need an expensive algorithm running. I'm not testing performance, just testing the practicality of the feature.

Because Fibonacci is a pure algorithm it'll always return the same value given the same input. Someone actually reading the unit tests to understand how it works won't know the difference between both calls. Here I'm enforcing that expensive will increment with every call, however, because it is cached, it won't.

And this is also why I won't make this change, I want to emphasize that the call is skipped. Adding that fancy code will just confuse the reader.

ljmf00 replied:
Sure, I can agree about the unnecessary complexity, but then please do the latter. To clarify, rename the function `fib` to something else.

iK4tsu replied:
done

ljmf00 commented on source/decorated.d:
```diff
@@ -50,10 +50,18 @@ mixin template decorated(string name, alias fun)
+        else return call;
     }
 
     mixin("alias "~name~" = _internal_helper_;");
```
Not really related to this change, but we should check conformance with identifier grammar token (something like `([_a-zA-Z][_a-zA-Z0-9]*`, possibly allowing extra spaces too) to avoid weird errors. What if some user adds a `;` by mistake? We can give better error messages and later integrate with https://github.com/dlang/dmd/pull/13511 .

iK4tsu replied:
see line 5

ljmf00 replied:
That is weak checking and can be seen as insecure, if someone introduces text with `-J`. e.g.: `justkidding; void foo() {} int fake` this can be fed to trick the compiler.

iK4tsu replied:
done

ljmf00 commented on source/decorated.d:
```diff
@@ -50,10 +50,18 @@ mixin template decorated(string name, alias fun)
-        else
-            return impl!(__traits(getAttributes, _internal_helper_))();
+        auto call = impl!(__traits(getAttributes, _internal_helper_))();
+        static if (isCallable!(typeof(call)))
```
If it is not callable, what is the purpose of the contrary branching (line no 64)?

iK4tsu replied:
I mean, I can rename it to callee

ljmf00 replied:
I'm just trying to understand the logic of the alternative branching. What if `call` is not callable, what line number 64 falls for? In another perspective: is line 64 ever covered? If yes, for what purpose/case?

iK4tsu replied:
If it's not callable it's a value. Depending on how the decorator is implemented by the user it can either return a function or a value.

```d
int boo(T)(T fun) { return 4; }

@boo
mixin decorated!("fun", { });

void main()
{
	assert(fun() == 4);
}
```

I want to replicate as much as I can from Python's implementation, however, we can't just assign any value to a variable. In Python `fun` would be an `int`, in D we cannot do it that way, so the decorator will always be a function. To "simulate" that behavior, I evaluate the calls and get the last result. If the result is a value I return it otherwise it's a closure and I run it. This prevents code like `fun()()` which is not fun, annoying, and unintuitive.

ljmf00 replied:
Ah, I didn't know that was the case in Python. That makes sense either way. Thanks for the explanation.

ljmf00 commented on source/decorated.d:
```diff
@@ -135,4 +143,38 @@ unittest
+
+    with (S.init)
+    {
+        assert(fib(46) == 1); // runs expensive stuff
+        assert(fib(45) == 2); // runs expensive stuff
+
+        assert(fib(46) == 1); // skips the expensive stuff
+        assert(fib(45) == 2); // skips the expensive stuff
```
nit: avoid repeating the comments
```suggestion
	// runs expensive stuff
        assert(fib(46) == 1); 
        assert(fib(45) == 2);

	// skips the expensive stuff
        assert(fib(46) == 1);
        assert(fib(45) == 2);
```

iK4tsu replied:
done

ljmf00 commented on source/decorated.d:
```diff
@@ -135,4 +143,38 @@ unittest
         assert(bar() == "D roks: Hello Decorated");
         assert(baz() == "D roks: Hello Decorated");
```
Not related with this PR too, but `roks` -> `rox`

iK4tsu replied:
done

iK4tsu committed:
refactor: rename test example from 'D roks' to 'D rox'

Signed-off-by: João Lourenço <jlourenco5691@gmail.com>

iK4tsu committed:
fix: add attributes to unit tests

Signed-off-by: João Lourenço <jlourenco5691@gmail.com>

iK4tsu committed:
fix: correctly check for a valid identifier

Currently the function does not allow unicode characters which is not
conformant with the D spec.

Moreover the verification is weak which can lead security issues. The
user can pass multiple tokens and the verification might be valid:

```d
static assert(isDecoratedName!"_; alias myNewAlias");
```

This is also fixed with the new algorithm.

Signed-off-by: João Lourenço <jlourenco5691@gmail.com>
